### ISSUE TITLE
Clarify that getSynchronizationSources() should return information even if the track has no sink (<video> tag)

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-pc#2240]
When [Chrome tried to update its getSynchronizationSources() implementation](https://bugs.chromium.org/p/webrtc/issues/detail?id=10545#c14) to match [the spec](https://w3c.github.io/webrtc-pc/#dom-rtcrtpreceiver-getsynchronizationsources) about when the information is updated...

> When an audio or video frame from one or more RTP packets is delivered to the RTCRtpReceiver's MediaStreamTrack, the user agent MUST queue a task to update the relevant information for the RTCRtpContributingSource and RTCRtpSynchronizationSource dictionaries based on the content of those packets. 

... it started failing the [getSynchronizationSources() tests](https://cs.chromium.org/chromium/src/third_party/blink/web_tests/external/wpt/webrtc/RTCRtpReceiver-getSynchronizationSources.https.html) because without the MediaStreamTrack having a sink, Chrome was performing optimizations not to decode frames, and so it wasn't updating the SSRC information.

The question is: Is this spec compliant?

> As stated in the conformance section, requirements phrased as algorithms may be implemented in any manner so long as the end result is equivalent. So, an implementation does not need to literally queue a task for every frame, as long as the end result is that within a single event loop task execution, all returned RTCRtpSynchronizationSource and RTCRtpContributingSource dictionaries for a particular RTCRtpReceiver contain information from a single point in the RTP stream.

I would expect the MediaStreamTrack to, at least conceptually, receive frames as long as we are receiving decodable packets, whether or not the track forwards those frames to a sink such as a &lt;video&gt; tag.

On the other hand, why decode packets if we're not using them?

So...
1. Should we add a note to the spec saying "if the MediaStreamTrack does not have any sinks, this step MAY be optimized away, so that the SSRC information is not updated"?
2. Or should getSSRC be clarified to make the packet information mandatory even if there are no sinks and no decoding happening?
3. Or should decoding be mandatory, leaving no ambiguity here?

**Comment by na-g** [comment]
@henbos, on a related note, while implementing this interface in Firefox a number of things occurred to me. This interface cuts across 3 layers, RTP, decoding, and presentation. Each layer has additional filters that will cause a packet not to be reported. That means that if one is interested in the data from the perspective of a particular layer (say RTP: is so-and-so delivering media to me?), one has to contend with the layers above it (i.e presentation: am I seeing media from so-and-so?). As long as there is this tight coupling, I can see this interface being revisited for new use cases, and new corner cases like the one above.

If there were an interface that reported the RTP data when it arrived, that could satisfy the use-cases where media isn't being presented. The rtpTimestamp + SSRC could correlate the two sets of data. If this were (one or more) callback(s) the browser could defer work until the callback was registered, and assured that each packet was observed. This would allow for option 2) without suppressing the availability of the RTP information.

On 3) I think that the presence of this interface should mandate decoding (as it used to when the audio level had to be decoded if it wasn't present).

On 1) does it make sense for this behavior to be implementation dependent? Is there a reason not to upgrade the MAY to MUST, ensuring that one still only gets source information for presented packets?

**Comment by chxg** [comment]
This seem to be a side-effect of having a "frame pulling" architecture vs. a "frame pushing" one and goes beyond the implementation details of `getSynchronizationSources()` itself.

Video uses the "frame pushing" approach in Chromium. RTP packets arrive. They get decoded. Frames are then "pushed" to the track along with a timestamp of when it should be rendered. This is why the video web tests continued to pass after moving the gathering of info for `GetSynchronizationSources()` from packet receive time to frame delivery time:

https://bugs.chromium.org/p/webrtc/issues/detail?id=10545#c8

Audio uses the "frame pulling" approach in Chromium. RTP packets arrive. They get queued up in the jitter buffer in NetEq. They remain in the jitter buffer until the sound card "pulls" a 10 ms audio frame by ultimately calling `NetEq::GetAudio()`. Without a proper sink, the call is never made and packets stay in the buffer indefinitely. Or more commonly, the packets get discarded when the jitter buffer reaches capacity and is flushed by subsequent `NetEq::InsertPacket()` calls. This why the audio web tests started to fail after attempting to move the gathering of info for `GetSynchronizationSources()` from packet receive time to frame delivery time:

https://bugs.chromium.org/p/webrtc/issues/detail?id=10545#c12

This behavior is also noticeable elsewhere. We can take a look at the WebRTC sample named "Peer connection as input to Web Audio":

https://webrtc.github.io/samples/src/content/peerconnection/webaudio-output/

Pressing "Start" followed by "Call" shows a visualization of the audio from the microphone. But if we go to "main.js" and remove the line that says `remoteVideo.srcObject = e.streams[0];` from `function gotRemoteStream(e)`, we suddenly get nothing with Chromium. Firefox works either way.

**Comment by jan-ivar** [comment]
> if we go to "main.js" and remove the line that says `remoteVideo.srcObject = e.streams[0];` from `function gotRemoteStream(e)`, we suddenly get nothing with Chromium.

This is a bug clearly, since web audio is a sink. I don't see anywhere that says web audio need only work for audio that is already played out through other means—kind of defeats the purpose of web audio.

**Comment by jan-ivar** [comment]
I'd suggest the same applies to `getSynchronizationSources()`. See audio bars next to speakers in a WebRTC call seems useful regardless of whether that audio is being outputted. Maybe even more so when it is not.

Browsers that want to optimize this, could look at whether the JS is using the `getSynchronizationSources()` API at all (I suspect very few do)—The API is synchronous, but is meant to be called in realtime, so I don't think anyone would fault the browser if the very first call didn't have the right info?

**Comment by henbos** [comment]
I'll clarify that it is expected to return this information whether or not sinks are attached in the form of a note. I think it is up to the implementation how often to decode, or even whether to decode, when there are no sinks, as long as it is not testably incorrect. What should be clear for interop is that if packets are being received getSynchronizationSources() should give me recent information, i.e. attaching a track to a dummy sink should not be a prerequisite to using this API.

**Comment by chxg** [comment]
I don't believe that it adds sufficient clarity to the API.

`getSynchronizationSources()` returning an empty dictionary (in cases without a sink) remain a valid interpretation for as long as an implementation is allowed to postpone the consumption of packets indefinitely.

For example:

1. Things are setup without a sink.
2. Packets are received and inserted into the jitter buffer.
3. Packets are not consumed since there's no sink.
4. `getSynchronizationSources()` is called.
5. Waits 20 seconds.
6. Sink is added.
7. Packets that were queued up are now consumed and turned into frames that gets delivered to the track.
8. `getSynchronizationSources()` is called.

If (3) is allowed then (4) must return an empty dictionary, or else (8) won't capture the info for the frames in (7) despite them being delivered within the last 10 seconds.

I think that the clarification/spec-change needs to be on the interaction between `RTCRtpReceiver` and its `MediaStreamTrack`. E.g. declaring that packets must be consumed, and treating frames as if they are delivered to the track, regardless of whether there's a sink or not.

An implementation may do certain invisible optimizations, such as not spending CPU cycles on the actual audio/video codec work, but "behaving as frames are not delivered to the track until a sink is added" doesn't seem to be one of them.

ps. Some of the stats in `GetStats()` like `RTCAudioReceiverStats.totalSamplesDuration` should be similarly affected by sink-less tracks since they too depend on audio frames being delivered to the track.

**Comment by henbos** [comment]
Yeah that's a good point, and getStats() is another API that exposes Chrome's current behavior when having no sink. Placing the note around the RTCRtpReceiver section is a good place since the requirement to consume has multiple implications.

**Comment by jan-ivar** [comment]
It seems like Chrome is picking up side-effects from trying to meet the new spec definition of when *"RTP packets is delivered to the RTCRtpReceiver's MediaStreamTrack"*. As I recall, that definition was a compromise time-point somewhere between RTP arrival and playout—it's still not a JS observable time-point, and not guaranteed to be the same in every implementation, so why?

Because it was as close as we could normatively get to playout without involving other specs. After all, the synchronous `getSynchronizationSources()` method's intent remains to help draw volume bars next to people talking (which can be derived solely from RTP) which meant A) doing so whether you hear them or not, and B) have the bars sync with them talking when you do.

Importantly: I don't think this change was meant to introduce dependencies on having sinks, nor to impact *getStats* in any way, whose mentions of audio AFAIK predates this change.

**Comment by chxg** [comment]
I don't believe that the changes at https://github.com/w3c/webrtc-pc/issues/2172 had any impact on this matter.

The original spec had language such as:

>  ... including the most recent time a packet that the source contributed to _was played out_...
>
> <...>
>
> When the first audio or video frame contained in an RTP packet is _delivered_ to the RTCRtpReceiver's MediaStreamTrack for _playout_...
>
> <...>
>
> The timestamp of type DOMHighResTimeStamp [HIGHRES-TIME], indicating the most recent time of _playout_ of media that arrived in an RTP packet originating from this source.

So the problem described here would still have been present with those definitions. One could even argue that the previous definitions might even be interpreted as demanding the returned dictionaries to be empty without a sink, since there's no "playout" without one.

**Comment by henbos** [comment]
Yeah the original definition was even more undefined, and all the problems discussed here would certainly have applied to a vague "playout" (what is playout if you have multiple sinks? when you have none?). I think the one we got now is good ("when delivered to the track") and is pretty well-defined.

The question is "when is decoding mandatory?". It could have been asked earlier, but due to previous implementation choices (not in-line with either the old or the new definition) the problem of not always decoding did not show its face until now.

The getStats side-issue is just another way to observe whether or not decoding is happening (both before or after these changes), and is not impacted by any getSSRCs changes, the point was that if we aren't clear about when decoding is mandatory then that is another way to observer it in JS even if the getSSRCs() API didn't exist.

Firefox didn't strictly speaking implement either definition since it was doing estimates and not actual measurements. The estimate is probably "good enough" to be used for the new implementation as well, even though it's not exactly what the spec says (either definition).

> Importantly: I don't think this change was meant to introduce dependencies on having sinks, nor to impact getStats in any way, whose mentions of audio AFAIK predates this change.

What increased the dependency on the sinks was trying to implement something closer to either version of the spec rather than estimating it. If we say that the browser MAY estimate the value, or specify whether or not decoding is mandatory, we're OK. I think we need to do this independently of which definition of getSSRC we use.

**Comment by jan-ivar** [comment]
@henbos Can you summarize if there's still an issue here? Do we want to add a note? Thanks.

**Comment by henbos** [comment]
I'll have to go through this again but basically I think we want to add a note.
